Controllers: Don't needlessly copy presets into user directory

Holzhaus opened the pull request:
~~**Based on PR: #2567**~~

For XML, we now check if the preset has been edited and only save it to the user directory in that case. Since we don't allow to edit JavaScript from within Mixxx, these are never copied anymore.

Now I can actually use Mixxx with an empty `~/.mixxx/controllers/` directory :tada:

Holzhaus committed:
controllers/controllermanager: Make use of the getPresetPaths() function

Holzhaus committed:
controllers/controllermanager: Rename function to sanitizeString

Holzhaus committed:
controllers/controllermanager: Rework device setup code

Holzhaus committed:
controllers: Preselect configured mapping in preferences if available

Holzhaus committed:
controllers/controllerpresetinfo: Remove unused header file

Holzhaus committed:
controllers/dlgprefcontroller: Separate user and system presets

Holzhaus committed:
controllers/dlgprefcontroller: Fix reset of combobox item on apply

Holzhaus committed:
controllers/controllermanager: Always write preset to config on exit

Holzhaus committed:
controllers/midi: Make MidiControllerPreset's mappings private

Holzhaus committed:
controllers: Only write preset to user directory if it's dirty

Holzhaus committed:
controllers/dlgprefcontroller: Don't import scripts into user directory

uklotzde commented:
The copying and duplicate devices in the list boxes with the same name cause so much confusion, even for me. This would be a huge improvement!

uklotzde commented:
I guess this should to be included in 2.3.0?

ronso0 commented:
I didn't built this, yet.
How's the workflow now if I want to alter the js file only?

Holzhaus commented:
> I didn't built this, yet.
> How's the workflow now if I want to alter the js file only?

I think you can just edit the file and save it in `~/.mixxx/controllers/`.

Holzhaus commented:
> I guess this should to be included in 2.3.0?

I wasn't sure since @ywwg was keen on reducing the scope of 2.3. If nobody encounters any problems during testing I'd like to get this merged ASAP though ;-)

Be-ing commented:
Thank you for working on this obnoxious UX issue.

> I guess this should to be included in 2.3.0?

Ehhh maybe we shouldn't rush this. We have enough on our plate for 2.3.

Holzhaus commented:
I'd propose this for 2.3 because it makes controller mapping development *much* easier. Right now, you can't really develop controller mappings from the git source tree, because Mixxx will always copy them to the user directory. On the wiki, we recommend symlinking `~/.mixxx/controllers` to `<gitroot>/res/controllers`, but the XML file will be copied anyway, so you need to change the controller mapping every time you restart mixxx.

You can still develop from the user directory, but then there is no way to use pre-commit hooks.

goddisignz commented:
Awesome! I recently started working on a component based mapping for the Pioneer DDJ-SX2. I rebased your PR onto my working branch and everything is going so much smoother now.
:sunglasses:

Holzhaus commented:
> Awesome! I recently started working on a component based mapping for the Pioneer DDJ-SX2. I rebased your PR onto my working branch and everything is going so much smoother now.

Great! I hope we can get this reviewed and merged before we branch off 2.3. This makes it possible to develop mappings inside the git repo and use the pre-commit hooks to ensure proper code style.

Holzhaus committed:
controllers/controllermanager: Rename sanitizeString and move to anon NS

Be-ing commented:
I think we have several use cases to consider here. Please clarify what you propose should be done in each if we merge this branch:
  * User using mapping shipped with Mixxx: select from combobox in preferences, mapping name stays selected
  * User using mapping from other source: copy XML & JS files to user settings folder, then select from combobox and the mapping name stays selected?
  * Developing a mapping: Run `mixxx --resourcePath /path-to-git-repo/res/`?

Be-ing commented:
Can you make the "Enabled" checkbox automatically check when a mapping is selected?

Be-ing commented on src/controllers/controllermanager.cpp:
```diff
@@ -218,24 +218,24 @@ void ControllerManager::slotSetUpDevices() {
-        if (!loadPreset(pController, pPreset)) {
-            // TODO(XXX) : auto load midi preset here.
+        QString presetFile = m_pConfig->getValueString(
+                ConfigKey("[ControllerPreset]", deviceName));
```
Woohoo this finally fixes [using multiple controllers of the same model](https://bugs.launchpad.net/mixxx/+bug/1522507)! I think @mevseme will be happy. :)

Be-ing replied:
From here it is a short step to automatically loading mappings instead of making the user select them.

Be-ing commented on src/controllers/controllermanager.cpp:
```diff
@@ -392,7 +392,17 @@ void ControllerManager::slotSavePresets(bool onlyActive) {
+                    << "Preset for device" << deviceName
+                    << "is not dirty, no need to save it to the user presets.";
+            continue;
+        }
```
Could you append " (edited)" to the name of the preset here?

Be-ing replied:
Maybe this isn't the place to do it. The code that handles the mapping editing tables should change the name so that " (edited)" doesn't get appended repeatedly when the user edits multiple times.

Be-ing replied:
Check if it's a user or system mapping. If it's a user mapping, don't append " (edited)"

Holzhaus replied:
Good idea. That way we don't end up with "My Mapping (edited) (edited) (edited)".

Be-ing commented:
Oh yay it doesn't scramble the order of elements in modified XML files :)

Be-ing commented on src/controllers/controllermanager.h:
```diff
@@ -85,7 +85,7 @@ class ControllerManager : public QObject {
 
-    static QString presetFilenameFromName(QString name) {
+    static QString sanitizeString(QString name) {
         return name.replace(" ", "_").replace("/", "_").replace("\\", "_");
```
It is a shame that Qt does not provide a solution for this common problem. I was just dealing with this the other day for #2618. Could you move this to `src/util/string.h`? Also rename the function to `sanitizeFileName`.

Be-ing replied:
I don't think this needs to replace spaces, does it? What does not support spaces in file names?

Holzhaus replied:
> I don't think this needs to replace spaces, does it? What does not support spaces in file names?

That was the main reason I renamed it. It's sometimes used for file names, but it's also uses for formatting the device name so that it can be used as a config key. I'll rename the function to `sanitizeDeviceName`.

Be-ing replied:
okay

Be-ing commented on src/controllers/controllermanager.cpp:
```diff
@@ -218,24 +218,24 @@ void ControllerManager::slotSetUpDevices() {
-                ControllerPresetFileHandler::loadPreset(
-                    presetBaseName + pController->presetExtension(),
-                    getPresetPaths(m_pConfig));
+        if (m_pConfig->getValueString(ConfigKey("[Controller]", deviceName)) != "1") {
```
Is this the state from the enabled checkbox? If so, please add a comment. It is not obvious from the code.

Holzhaus replied:
Yes it is. I'll add a comment.

Be-ing commented on src/controllers/controllermanager.cpp:
```diff
@@ -392,15 +392,15 @@ void ControllerManager::slotSavePresets(bool onlyActive) {
-                       << name << "to" << presetPath;
+                       << deviceName << "to" << presetPath;
         }
+        m_pConfig->set(ConfigKey("[ControllerPreset]", deviceName), presetPath);
```
Is this a new ConfigKey? If so, I'd like to get away from this awkward "preset" terminology. No other DJ software uses it, everyone else calls them "mappings". I think "preset" came from a mistaken concept that more than a small fraction of users care to edit their controller mappings but actually they just want them to work. I'm not going to insist on renaming all the preset related code in this PR, but let's not get stuck with this ConfigKey for backward compatibility.

Holzhaus replied:
No, it exists since at least 1.12: https://github.com/mixxxdj/mixxx/commit/e24a59a25ed2dd479dbcbf8bbc7eacf1eb89ac13#diff-2394e2fc81b545b16e24ea386931aaefR310-R317

Holzhaus replied:
We could rename, but then users need to setup their controllers again.

Be-ing replied:
Okay, let's leave it then. Thanks for checking.

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -239,25 +239,45 @@ void DlgPrefController::enumeratePresets() {
     m_ui.comboBoxPreset->addItem("...");
 
-    // Ask the controller manager for a list of applicable presets
-    QSharedPointer<PresetInfoEnumerator> pie =
```
There really was a variable named "pie" :rofl:

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -239,25 +239,45 @@ void DlgPrefController::enumeratePresets() {
+    // Ask the controller manager for a list of applicable user presets
+    QSharedPointer<PresetInfoEnumerator> userPresetEnumerator =
+            m_pControllerManager->getMainThreadUserPresetEnumerator();
+    // Check if enumerator is ready. Should be rare. We will re-enumerate on
```
```suggestion
    // Check if enumerator is ready. Should be rare that isn't. We will re-enumerate on
```

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -239,25 +239,45 @@ void DlgPrefController::enumeratePresets() {
+    // Check if enumerator is ready. Should be rare. We will re-enumerate on
+    // the next open of the preferences.
+    if (!userPresetEnumerator.isNull()) {
+        // Making the list of presets in the alphabetical order
```
```suggestion
        // Sort list alphabetically
```

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -239,25 +239,45 @@ void DlgPrefController::enumeratePresets() {
-    QList<PresetInfo> presets = pie->getPresetsByExtension(
-        m_pController->presetExtension());
+    // Insert a separator between user presets (+ dummy item) and system presets
+    m_ui.comboBoxPreset->insertSeparator(m_ui.comboBoxPreset->count());
```
Great idea :+1:

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -239,25 +239,45 @@ void DlgPrefController::enumeratePresets() {
+            m_pControllerManager->getMainThreadSystemPresetEnumerator();
+    // Check if enumerator is ready. Should be rare. We will re-enumerate on
+    // the next open of the preferences.
+    if (!systemPresetEnumerator.isNull()) {
+        // Making the list of presets in the alphabetical order
+        QList<PresetInfo> systemPresets = systemPresetEnumerator->getPresetsByExtension(
+                m_pController->presetExtension());
+
+        for (const PresetInfo& preset : systemPresets) {
+            m_ui.comboBoxPreset->addItem(preset.getName(), preset.getPath());
+            if (m_pController->matchPreset(preset)) {
+                match = preset;
+            }
```
factor out this duplicated code to a private method

Holzhaus replied:
Done.

Be-ing replied:
Where? Did you forget to push?

Be-ing commented on src/controllers/controllermanager.cpp:
```diff
@@ -392,7 +392,17 @@ void ControllerManager::slotSavePresets(bool onlyActive) {
+        DEBUG_ASSERT(!pPreset);
+
         QString deviceName = sanitizeString(pController->getName());
+        if (!pPreset->isDirty()) {
```
Should this be a `VERIFY_OR_DEBUG_ASSERT`?

Holzhaus replied:
No, this is always called even if the mapping was not changed right now.

Be-ing replied:
okay

Holzhaus committed:
controllers/controllermanager: Fix assertion if no preset is configured

Holzhaus committed:
controllers/controllermanager: Use qDebug() instead of qWarning()

Holzhaus committed:
controllers/dlgprefcontroller: Move shared enumeration code to method

Holzhaus committed:
controllers/dlgprefcontroller: Improve some comments

Be-ing commented:
Somehow I got this into a state where it isn't showing any mappings from my system folder in the combobox. I tried running with a fresh `--settingsPath` but it is still happening...

Holzhaus committed:
controllers/controllermanager: Improve device setup code

Holzhaus committed:
controllers/controllermanager: Remove unused variable

Be-ing commented:
> Somehow I got this into a state where it isn't showing any mappings from my system folder in the combobox. I tried running with a fresh --settingsPath but it is still happening...

Haha, nevermind that was my fault. I accidentally deleted the "controllers" folder in my --resourcePath

Be-ing commented:
Please modify PresetInfoEnumerator so it works with a single QString path instead of a QStringList.

Be-ing commented:
Can you insert dummy items into the mapping combobox as headers for each section, "Custom mappings" (or "User mappings"?) and "Mixxx bundled mappings" (open to wording suggestions)?

Be-ing commented:
I was concerned this would break my development workflow with symlinking ~/.mixxx/controllers to the res/controllers folder in a separate git worktree from my C++ development, but it actually works fine.

Be-ing commented:
I tested with an empty ~/.mixxx/controllers except for one XML & JS file that depends on midi-components-0.0.js and it worked fine loading that. Components was loaded from the resource directory as expected. :+1:

Be-ing commented:
Thanks for putting in the work to fix this clunky UX. Originally there was no issue with mappings appearing in the list twice after selecting a bundled mapping from the resource folder. I broke that with one of my first C++ PRs #654, which was a simple one line hack that didn't really account for all the implications of reading mappings from the user folder.

Holzhaus committed:
controllers/dlgprefcontroller: Add icon to user controller presets

Holzhaus commented:
> Can you insert dummy items into the mapping combobox as headers for each section, "Custom mappings" (or "User mappings"?) and "Mixxx bundled mappings" (open to wording suggestions)?

I thought about that too, but apparently there's no way to make it non-selectable, which is confusing. With this PR, the user preset directory only contains mappings that have been edited and those that have been downloaded and manually copied there, so unless you symlink it to the `res/controllers` it should just be a few mappings (if at all). For these custom mappings I added an icon.

Be-ing commented:
> For these custom mappings I added an icon.

That works too.

Be-ing commented:
How about using the Mixxx logo as the icon for the bundled mappings?

Holzhaus committed:
controllers/controllerpresetinfoenumerator: Add constructor for single path

Holzhaus committed:
controllers/dlgprefcontroller: Show Mixxx icon for system presets

Be-ing commented:
build fails
```
be@localhost ~/s/mixxx ((32ba464b…))> 
cmake --build ~/sw/mixxx/cbuild --target install --parallel 4
[  1%] Built target mixxx-xwax
[  2%] Built target MP3GuessEnc
[  2%] Built target FpClassify
[  3%] Built target libebur128
gmake[2]: *** No rule to make target '../res/images/ic_mixxx_symbolic.svg', needed by 'mixxx-qrc_autogen/PNK5WDWK6L/qrc_mixxx.cpp'.  Stop.
gmake[1]: *** [CMakeFiles/Makefile2:1406: CMakeFiles/mixxx-qrc.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
[  3%] Built target mixxx-hidapi
[  4%] Built target fidlib
[  4%] Built target PortAudioRingBuffer
gmake: *** [Makefile:163: all] Error 2
be@localhost ~/s/mixxx ((32ba464b…)) [2]>
```

Be-ing commented:
Looks like you forgot to commit the new file?

Be-ing commented on src/controllers/controllerpresetinfoenumerator.cpp:
```diff
@@ -29,6 +29,10 @@ bool presetInfoNameComparator(const PresetInfo &a, const PresetInfo &b) {
+        : PresetInfoEnumerator(QList<QString>{searchPath}) {
+}
+
 PresetInfoEnumerator::PresetInfoEnumerator(const QStringList& searchPaths)
```
remove this old constructor

Holzhaus replied:
I thought I leave it in so that we can easily add additional directories later, for example to add support for the XDG Directory Spec while maintaining backward compatibility with mappings in `~/.mixxx`. I you insist, I can also remove that contructor.

Be-ing replied:
Okay, I didn't realize you had future plans for it.

Holzhaus replied:
Yes, but not for 2.3 :-)

Holzhaus commented:
> Please modify PresetInfoEnumerator so it works with a single QString path instead of a QStringList.

Done.

> How about using the Mixxx logo as the icon for the bundled mappings?

Good idea. There was no black/white symbolic icon, so I had come up with something in Inkscape myself.

Holzhaus commented:
> Looks like you forgot to commit the new file?

Jup, sorry.

Be-ing commented:
> There was no black/white symbolic icon, so I had come up with something in Inkscape myself.

It looks fine :+1:

Be-ing commented:
> Can you make the "Enabled" checkbox automatically check when a mapping is selected?

Don't forget this.

Holzhaus committed:
controllers/dlgprefcontroller: Disable checkbox if no preset selected

Holzhaus committed:
controllers/dlgprefcontroller: Rename "..." to "No Preset" and add icon

Holzhaus committed:
controllers/controllermanager: Fix typo in debug message

ronso0 commented:
Very nice!
A minor issue (for me): having the User and Mixxx icon on every mapping feels redundant and just adds noise.
Can the Mixxx icon be attached to the separator? (+ icon & separator for user mappings)

Be-ing commented:
I'm fine with it how it is.

Holzhaus commented:
> > Can you make the "Enabled" checkbox automatically check when a mapping is selected?
> 
> Don't forget this.

Yes, I'll take care of that tomorrow. 

> A minor issue (for me): having the User and Mixxx icon on every mapping feels redundant and just adds noise.
> Can the Mixxx icon be attached to the separator? (+ icon & separator for user mappings)

No, that"s not possible (AFAIK). Adding a title for a separator isn't possible either. We can remove the icon for system presets, and only use an icon for user presets (of which you usually only have a few). That was the case before I added the Mixxx icon to system presets.

Holzhaus commented:
I also found a bug with the dirty flag when editing a mapping via the input/output mapping table. I'll investigate tomorrow.

Be-ing commented:
Please add documentation to the header files of the classes you are modifying.

Holzhaus committed:
Merge branch 'master' of github.com:mixxxdj/mixxx into controller-fixes-part2

Holzhaus committed:
controllers/midi/midimessage: Use desc in  MidiInputMapping == operator

Holzhaus committed:
controllers/midi/midimessage: Fix MidiInputMapping == operator

Holzhaus committed:
controllers/controllermanager: Add "(edited)" suffix to changed presets

Holzhaus committed:
controllers: Add some comments to undocumented classes

Holzhaus committed:
controllers/controllermanager: Remove obsolete importScript method

ronso0 commented:
> > A minor issue (for me): having the User and Mixxx icon on every mapping feels redundant and just adds noise.
> > Can the Mixxx icon be attached to the separator? (+ icon & separator for user mappings)
> 
> No, that"s not possible (AFAIK). Adding a title for a separator isn't possible either. We could remove the icon for system presets, and only use an icon for user presets (of which you usually only have a few). That was the case before I added the Mixxx icon to system presets. Not sure if that's better.

Okay, than we better leave it as it is. Thx!

Holzhaus committed:
controllers: Always load scripts from mapping dir (with system as fallback)

Holzhaus committed:
controllers/dlgprefcontroller: Add "(built-in)" to system scripts

Holzhaus commented:
> I also found a bug with the dirty flag when editing a mapping via the input/output mapping table. I'll investigate tomorrow.

Fixed. The `==` operator for midi input mappings just ignored the mapping description to make writing the unit tests easier :-)

> Please add documentation to the header files of the classes you are modifying.

Done.

Holzhaus commented:
By the way I found a pretty major bug with loading mapping scripts. The scripts were *always* searched in the user directory first, then in the system directory.

That means that if you put a modified `midi-components.js` into your user directory, it would always be used, even for built-in mappings. If a user downloaded a mapping that comes with a custom `midi-components.js`, that could potentially break all built-in mappings.

I changed it so that the mapping's directory (i.e. where the XML file is located) is always used first, then it falls back to the system directory (if mapping directory != system directory). That way you can still override `midi-components.js` from your user directory, but you need to copy the mapping XML to your user directory, too.

Holzhaus commented:
I'll try to centralize the code that determines the script file paths a bit, then this is ready.

Holzhaus commented:
By the way, I updated the script list to show if a script from from the system directory. This is how I noticed the issue :wink: 

![2020-04-07-145235_971x575_scrot](https://user-images.githubusercontent.com/1834516/78671412-8e614180-78df-11ea-8dfb-60851778c586.png)

ronso0 commented:
With this branch I noticed I had a bunch of mappings in my user folder from testing #2394 (script links). From the GUI I opened that folder and removed obsolete mapppings.
It would be nice if the mapping selector would be re-populated correctly, either **a)** when switching to the Controller Setup tab or **b)** when opening the mapping selector.

ronso0 commented:
I played around with this in --developer mode with MIDI Throug Port-0.
I hope this testing procedure was correct, otherwise disregard my findings.

* Pref > Controllers > MIDI Throug Port-0
* selected `Akai MPD24`
* Output Mappping tab > added a random output mapping > Apply
* Controller Setup tab: not showing `(edited)` suffix
* Apply, Close
* open Pref > Controllers > MIDI Throug Port-0
mapping selector: `(user) Reloop Terminal ...`
mapping name: `Akai MPD24 (edited)`
no Akai MPD24 mapping in user folder, but Midi_Through_Port-0.midi.xml was overwritten
* close/restart Mixxx
* open Pref > Controllers > MIDI Throug Port-0
mapping selector: `(user) Akai MPD24 (edited) (edited)`
mapping name: `Akai MPD24 (edited) (edited)`

Be-ing commented on src/controllers/controllermanager.cpp:
```diff
@@ -415,6 +415,12 @@ void ControllerManager::slotSavePresets(bool onlyActive) {
         }
 
+        if (pPreset->filePath().startsWith(resourcePresetsPath(m_pConfig))) {
+            pPreset->setName(QString(tr("%1 (edited)")).arg(pPreset->name()));
```
```suggestion
            pPreset->setName(pPreset->name + " (" + tr("edited") + ")");
```
Let's not bother translators with the formatting.

Be-ing commented:
> By the way I found a pretty major bug with loading mapping scripts. The scripts were always searched in the user directory first, then in the system directory.

I found that too but I thought it was okay because Mixxx no longer copies to the user folder automatically. However I am glad you improved it.

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -224,16 +224,27 @@ QString DlgPrefController::presetScriptFileLinks(const ControllerPresetPointer p
+
+            qDebug() << "path" << path << systemPresetPath;
+            if (path.startsWith(systemPresetPath)) {
+                scriptFileLink = QString(tr("%1 (built-in)")).arg(scriptFileLink);
```
Nice, this is helpful.

Be-ing commented:
I confirm @ronso0's findings with the XML editing.

Holzhaus committed:
controllers: Move script file search code into ControllerPreset class

Holzhaus commented:
> I confirm @ronso0's findings with the XML editing.

I"ll investigate tomorrow. I spent way too much time figuring out how to fix the script search path issue, but IMHO it was necessary because it could prevent users that update from an older Mixxx version from benefiting from updated mappings just because they have an old version of midi-components rotting in their user directory.

Be-ing commented on src/controllers/controllerpreset.h:
```diff
@@ -43,16 +45,38 @@ class ControllerPreset {
+        QDir path(fallbackPath);
+        QList<ScriptFileInfo> scriptsWithFallbackPath;
+        for (const ScriptFileInfo& script : m_scripts) {
+            qDebug() << "script" << script.file;
```
This is not a useful debug message to leave in. Either add more informative text or remove this line.

Holzhaus replied:
Oops, leftover from debugging. I'll remove it.

Be-ing commented on src/controllers/controllerpreset.h:
```diff
@@ -43,16 +45,38 @@ class ControllerPreset {
+        for (const ScriptFileInfo& script : m_scripts) {
+            qDebug() << "script" << script.file;
+            if (script.file.exists()) {
+                scriptsWithFallbackPath << script;
```
The variable name `scriptsWithFallbackPath` is confusing here. At this line, it is scripts without fallback paths getting added. How about renaming the QList to `scriptsIncludingFallbacks`?

Be-ing commented on src/controllers/controllerpreset.h:
```diff
@@ -43,16 +45,38 @@ class ControllerPreset {
         setDirty(true);
     }
 
+    QList<ScriptFileInfo> getScriptFiles(const QString& fallbackPath = QString()) const {
```
I am doubting that this function should exist. Why not do all this in `addScriptFile`? Doing all this every time the ControllerPreset's scripts are used is superfluous.

Be-ing replied:
This will take more refactoring, but I think it will be worth it.

Holzhaus replied:
Done.

Be-ing commented on src/controllers/controllerengine.cpp:
```diff
@@ -933,35 +928,22 @@ void ControllerEngine::trigger(QString group, QString name) {
-        filename = scriptName;
-        input.setFileName(filename);
-    } else {
-        for (const QString& scriptPath : scriptPaths) {
```
Good idea to move this out of ControllerEngine. It doesn't belong here.

Be-ing commented on src/controllers/controllerengine.cpp:
```diff
@@ -933,35 +928,22 @@ void ControllerEngine::trigger(QString group, QString name) {
-                break;
-            }
-        }
+    if (!scriptFile.exists()) {
```
VERIFY_OR_DEBUG_ASSERT? This should be checked before it gets to ControllerEngine, right?

Holzhaus replied:
Unsure. It wasn't checked before. To make this a debug assert we'll have to add code that filter out files that don't exist to `ControllerPreset`.

Be-ing replied:
I think we should do that.

Be-ing replied:
On second thought, maybe not. If we did that, the preferences couldn't show that a file is missing.

Be-ing commented on src/controllers/controllerpresetfilehandler.cpp:
```diff
@@ -198,7 +196,7 @@ QDomDocument ControllerPresetFileHandler::buildRootWithScripts(const ControllerP
     controller.appendChild(scriptFiles);
 
-    foreach (const ControllerPreset::ScriptFileInfo& script, preset.scripts) {
+    foreach (const ControllerPreset::ScriptFileInfo& script, preset.getScriptFiles()) {
```
Replace legacy Qt `foreach (foo, bar)` macro with standard C++ `for (foo : bar)`

Be-ing commented on src/controllers/controllerengine.cpp:
```diff
@@ -933,35 +928,22 @@ void ControllerEngine::trigger(QString group, QString name) {
         return false;
     }
 
-    QString filename = "";
```
These changes to ControllerEngine are likely creating conflicts with #1795. Could you take care of those after this is merged? I don't think it'd be fair to ask @ferranpujolcamins to keep fixing the the merge conflicts when that branch has been waiting for a long time.

Holzhaus committed:
controllers/controllerpreset: Remove leftover qDebug() statement

Holzhaus committed:
controllers/controllerpresetfilehandler: Use for loop instead of foreach

Holzhaus committed:
controllers: Move controller script file handling to FileHandler classes

Be-ing commented on src/controllers/controllerpresetfilehandler.h:
```diff
@@ -16,15 +16,17 @@ class ControllerPresetFileHandler {
     ControllerPresetFileHandler() {};
```
I struggled to understand the purpose of this class and why it is separate from ControllerPreset. The documentation at the top of the file was only a little helpful. After looking around the code, it seems this class's role is to serialize/deserialize ControllerPreset objects to/from XML and save/load those files to/from the filesystem. Am I understanding that right?

Holzhaus replied:
Yes. Added a comment.

Be-ing commented on src/controllers/controllerpresetfilehandler.cpp:
```diff
@@ -13,56 +13,53 @@
-                 << "in any preset path.";
+ControllerPresetPointer ControllerPresetFileHandler::loadPreset(
+        const QFileInfo& presetFile, const QDir& systemPresetsPath) {
+    VERIFY_OR_DEBUG_ASSERT(presetFile.exists() && presetFile.isReadable()) {
```
I don't think this should be a debug assertion. This function is taking in untrusted data from the filesystem that has not been validated yet.

Holzhaus replied:
Correct. I was originally planning to check this beforehand, but that would still suffer from TOCTOU issues...

Be-ing commented on src/controllers/controllerpresetfilehandler.cpp:
```diff
@@ -13,56 +13,53 @@
-
-    ControllerPresetFileHandler* pHandler = NULL;
-    if (scriptPath.endsWith(MIDI_PRESET_EXTENSION, Qt::CaseInsensitive)) {
+    ControllerPresetFileHandler* pHandler = nullptr;
```
I don't think a pointer needs to be explicitly initialized to nullptr?

Holzhaus replied:
I think it does. It's not always assigned a value below.

Be-ing replied:
okay

Be-ing commented on src/controllers/controllerpresetfilehandler.cpp:
```diff
@@ -13,56 +13,53 @@
+                       HID_PRESET_EXTENSION, Qt::CaseInsensitive) ||
+            presetFile.fileName().endsWith(
+                    BULK_PRESET_EXTENSION, Qt::CaseInsensitive)) {
         pHandler = new HidControllerPresetFileHandler();
```
It's confusing that this class is called `HidControllerPresetFileHandler` but here it is also being used for bulk controllers.

Holzhaus replied:
Yes, but I don't want to do too much refactoring tbh.

Be-ing replied:
That's fine, we don't have to solve this now.

Be-ing commented on src/controllers/controllermanager.cpp:
```diff
@@ -241,9 +242,16 @@ void ControllerManager::slotSetUpDevices() {
+                 << "in paths:" << presetPaths.join(",");
+        QString presetFilePath = ControllerManager::getAbsolutePath(presetFile, presetPaths);
+        if (presetFilePath.isEmpty()) {
+            qDebug() << "Could not find" << presetFilePath << "in any preset path.";
```
Should we popup a dialog telling the user about this?

Be-ing replied:
ping

Be-ing commented on src/controllers/controller.h:
```diff
@@ -83,7 +83,7 @@ class Controller : public QObject, ConstControllerPresetVisitor {
 
     // Initializes the controller engine and returns whether it was successful.
-    virtual bool applyPreset(const QString& fallbackScriptPath, bool initializeScripts);
+    virtual bool applyPreset(bool initializeScripts);
```
```suggestion
   // The initializeScripts bool is only to bypass the script related code for tests
    virtual bool applyPreset(bool initializeScripts = true);
```

Be-ing commented on src/controllers/controllermanager.cpp:
```diff
@@ -219,6 +219,7 @@ void ControllerManager::slotSetUpDevices() {
     QList<Controller*> deviceList = getControllerList(false, true);
+    QStringList presetPaths(getPresetPaths(m_pConfig));
 
     foreach (Controller* pController, deviceList) {
```
```suggestion
    for (Controller* pController : deviceList) {
```

Be-ing commented on src/controllers/controllerpresetfilehandler.cpp:
```diff
@@ -13,56 +13,53 @@
-        qDebug() << "Preset" << scriptPath << "has an unrecognized extension.";
+    if (pHandler == nullptr) {
+        qDebug() << "Preset" << presetFile.absoluteFilePath()
+                 << "has an unrecognized extension.";
```
Popup a QMessageBox?

Be-ing commented on src/controllers/controllerpresetfilehandler.cpp:
```diff
@@ -111,16 +110,27 @@ void ControllerPresetFileHandler::addScriptFilesToPreset(
 
     // Default currently required file
-    preset->addScriptFile(REQUIRED_SCRIPT_FILE, "", true);
+    preset->addScriptFile(REQUIRED_SCRIPT_FILE,
+            "",
+            QFileInfo(systemPresetsPath.absoluteFilePath(REQUIRED_SCRIPT_FILE)),
+            true);
```
This prevents loading a modified common-controller-scripts.js from the user folder. I think we should handle this the same way as other scripts (preset path first, then fallback to system), just do it automatically without having to specify it explicitly in the preset.

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -390,7 +390,7 @@ void DlgPrefController::slotLoadPreset(int chosenIndex) {
     presetDirs.append(presetFileInfo.canonicalPath());
```
This code is now unused. Remove it.

Be-ing commented:
Thanks again for working on this. This area of the code hasn't gotten a lot of attention in many years. Clearly there is a lot of room for improvement. Let's try to not get carried away rewriting everything at this time, but I suspect we'll end up doing some major refactoring for 2.4.

Holzhaus committed:
controllers: Make applyPreset initializeScripts param optional

Holzhaus committed:
controllers/controllermanager: Replace foreach for for loops

Holzhaus committed:
controllers/controllerpresetfilehandler: Improve script file search code

Holzhaus committed:
controllers/controllerpresetfilehandler: Remove wrong debug assertion

Holzhaus committed:
controllers/controllerpresetfilehandler: Improve comments

Holzhaus committed:
controllers/dlgprefcontroller: Remove unused  presetDir code

Holzhaus committed:
controllers/controllerengine: Fix warnings for older Qt5 versions

Holzhaus committed:
controllers: Move savePreset() from Controller into Preset classes

Be-ing commented on src/controllers/controllerpresetfilehandler.h:
```diff
@@ -1,12 +1,17 @@
+* finding the script files that belong to a preset in the file system.
+*
+* Subclasses can implement the private load function to add support for XML
+* elements that are only useful for certain mapping types.
```
Thank you, this is better documentation :+1:

Be-ing commented on CMakeLists.txt:
```diff
@@ -217,6 +217,7 @@ add_library(mixxx-lib STATIC EXCLUDE_FROM_ALL
   src/controllers/keyboard/keyboardeventfilter.cpp
   src/controllers/learningutils.cpp
   src/controllers/midi/midicontroller.cpp
+  src/controllers/midi/midicontrollerpreset.cpp
```
Also update build/depends.py until we remove SCons support

Holzhaus replied:
Whoops, sorry. I'll fixup the commit.

Holzhaus committed:
controllers: Move preset saving from ControllerManger to Preferences

Holzhaus committed:
controllers/dlgprefcontroller: Reload user presets after saving

Holzhaus committed:
controllers/dlgprefcontroller: Remove obsolete method

Holzhaus committed:
controllers/dlgprefcontroller: Fix dirty detection in slotApply()

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -388,7 +388,7 @@ void DlgPrefController::slotApply() {
 
     // If no changes were made, do nothing
-    if (!isDirty() && (!m_pPreset || m_pPreset->isDirty())) {
+    if (!(isDirty() || (m_pPreset && m_pPreset->isDirty()))) {
```
Why do both DlgPrefController and ControllerPreset have dirty states?

Holzhaus replied:
The controller preset might be dirty because one of it's XML attributes or mappings might have changed.

DlgPrefController can be dirty when the enabled button was checked/unchecked or a different controller preset has been selected.

Be-ing replied:
`m_bDirty` is never set to true in DlgPrefController. Please remove it.

Holzhaus replied:
it is set via `setDirty()` on checkbox changes and when the mapping combobox is used:
https://github.com/mixxxdj/mixxx/blob/420ceb91b45fcfedb0e4c7629362ec43677b033c/src/controllers/dlgprefcontroller.cpp#L63
https://github.com/mixxxdj/mixxx/blob/420ceb91b45fcfedb0e4c7629362ec43677b033c/src/controllers/dlgprefcontroller.cpp#L432

Be-ing replied:
Oh it's set indirectly with the setDirty function. But that is only used when showing the mapping wizard, which IIUC changes the preset and marks it dirty anyway, and when the selected mapping in the combobox has changed. Does DlgPrefController even need to track that state with a bool member? How about explicitly comparing the loaded preset to the combobox selected preset in slotApply instead of complicating it with this "dirty" state?

Be-ing commented:
"Move preset saving from ControllerManger to Preferences"

Can you clarify your reasoning for this refactoring?

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -123,7 +134,7 @@ void DlgPrefController::showLearningWizard() {
 
     // After this point we consider the mapping wizard as dirtying the preset.
-    slotDirty();
+    setDirty(true);
```
I am confused here. Is this dirtying DlgPrefController or the preset? It is not obvious because they both get called "dirty".

Holzhaus replied:
dirtying the preset would be `m_pPreset->setDirty(true)` but i can add a comment

Be-ing replied:
I think a better solution would be to not use the word "dirty" and use words that convey what is happening.

Holzhaus replied:
Huh, no wonder you were confused: The line is wrong there. I removed it. The dirty state will be correctly set on the preset itself if anything has been learned by the wizard.

Holzhaus committed:
controllers/dlgprefcontroller: Reset dialog dirty state on apply

Holzhaus committed:
controllers/dlgprefcontroller: Remove wrong dirty state change

Holzhaus committed:
controllers/dlgprefcontroller: Add acomment

Holzhaus commented:
> "Move preset saving from ControllerManger to Preferences"
> 
> Can you clarify your reasoning for this refactoring?

The Controller Manager is responsible for managing controllers, not presets. I don't think that it should be responsible for saving presets. The only location where you can actually change presets is inside the DlgPrefController page, so it should also be saved there.

Also this makes it a lot easier to select the new saved preset in the checkbox, because we now know the filename. Otherwise we'd have to add an additional feedback path from controllermanager to the preference page.

Holzhaus commented:
The Controller page now has a working (!) apply button. Before, mappings would be applied instantly when selecting it on the combobox. Please test.

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -303,94 +368,115 @@ void DlgPrefController::slotUpdate() {
+        // User picked a preset
+        m_ui.chkEnabledDevice->setEnabled(true);
+
+        presetPath = m_ui.comboBoxPreset->itemData(chosenIndex).toString();
```
set the Enabled checkbox to enabled

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -303,94 +368,115 @@ void DlgPrefController::slotUpdate() {
+        m_ui.chkEnabledDevice->setEnabled(false);
+    } else {
+        // User picked a preset
+        m_ui.chkEnabledDevice->setEnabled(true);
```
add
`m_ui.chkEnabledDevice->setChecked(true);`

Holzhaus replied:
Hmm, so that selecting a preset means activating the controller? In that case we can remove the enabled checkbox completely and just use "no preset" for disabled controllers.

Be-ing replied:
Is there a use case for keeping a mapping selected but disabling the controller? The only time I have needed to disable and re-enable a controller was when the USB cable came unplugged during a set.

Holzhaus replied:
Wait, this works? I always restarted mixxx in that case.

Be-ing replied:
Indeed, it does. It's not obvious at all though.

Be-ing replied:
You have to manually disable the checkbox then re-enable it

Be-ing replied:
As a stopgap until we support controller hotplug, how about getting rid of the Enabled checkbox and treating every controller with a selected preset as enabled. For the reconnection edge case, we can add a Reconnect button that does what unchecking and rechecking the Enabled checkbox does now.

Holzhaus replied:
Hmm, how about we leave it as-is for 2.3? Users that upgrade from 2.2 will have a preset defined for all controllers because Mixxx was writing custom presets everywhere. This means that all controllers would be active. During development I made a little mistake that also activated all controllers - this made both my keyboard and my mouse unresponsive since they are also HID devices. I had to restart X11 via SSH to make my computer usable again :-)

Be-ing replied:
Oof, yeah. I was wondering if those legacy files laying in the user folders would have consequences. They sure would if we automatically enabled them. I still think it would help to add `m_ui.chkEnabledDevice->setChecked(true);` here in slotPresetSelected so users don't have to think about that checkbox.

Be-ing replied:
I think we can do this without automatically activating those legacy presets in the user folder. Here is what a fresh mixxx.cfg looks like from master after activating one of my Xone K2s:
```
Babyface_Pro_(70785713)_MIDI_1 /home/be/.mixxx/controllers/Babyface_Pro_(70785713)_MIDI_1.midi.xml
Babyface_Pro_(70785713)_MIDI_2 /home/be/.mixxx/controllers/Babyface_Pro_(70785713)_MIDI_2.midi.xml
XONE:K2_MIDI_1 /home/be/.mixxx/controllers/Allen and Heath Xone K2.midi.xml

[Controller]
XONE:K2_MIDI_1 1
```
I propose still relying on the `[Controller]` ConfigKey to determine which controllers to enable. Then automatically enable that old ConfigKey when a preset is applied. This would allow us to get rid of the Enabled checkbox.

Holzhaus replied:
Ok, this sounds like it could work. However, let's do this in a separate PR, this one is already big enough and I have the impression that PRs always take a few days without any new commits before they get merged here :wink:

Be-ing replied:
Okay

Be-ing commented:
The buttons to add and remove entries from the Input Mapping and Output Mapping tables have disappeared??

Holzhaus commented:
> The buttons to add and remove entries from the Input Mapping and Output Mapping tables have disappeared??

No, I didn't touch them and they are still there.

daschuer commented on src/controllers/controller.h:
```diff
@@ -82,8 +82,14 @@ class Controller : public QObject, ConstControllerPresetVisitor {
+     *  @param initializeScripts Can be set to false to skip script
+     *  initialization for unit tests.
+     *  @return Returns whether it was successful.
+     */
```
We use no block comments inside Mixxx

Be-ing commented:
master:
![Screenshot from 2020-04-08 15-08-15](https://user-images.githubusercontent.com/9455094/78830804-29285180-79ae-11ea-9435-69926f1b01a6.png)

this branch:
![Screenshot from 2020-04-08 15-31-28](https://user-images.githubusercontent.com/9455094/78830821-2f1e3280-79ae-11ea-8977-265e9fe44ffd.png)

Add, Remove, and Clear All buttons have disappeared

Holzhaus commented:
Is that a scrollbar on the right side of the window?

Be-ing commented:
Ah, they are hidden by the vertical scrollbar. How did that happen?

Holzhaus commented:
I have no idea. I didn't touch the UI files. I just built this on my laptop, this is how it looks for me:

![2020-04-08-22-39-52_1366x768](https://user-images.githubusercontent.com/1834516/78831488-0156df00-79ea-11ea-9a92-650d7a9702ee.png)

Be-ing commented:
I'm doing a git bisect...

daschuer reviewed (changes requested):
Some more comments

Be-ing commented:
After some more testing, I can reproduce the vertical scrollbar issue with the Inputs/Outputs tabs with master, so it is not a regression introduced by this branch and there is no need to fix it here. Maybe @ronso0 has some idea about it?

Holzhaus committed:
controllers/controllermanager: Return QFileInfo in findPresetFile()

Holzhaus commented on src/controllers/controllerinputmappingtablemodel.h:
```diff
@@ -9,6 +9,10 @@
+/** Table Model for the "Inputs" table view in the preferences dialog.
+ *
+ *  This allows editing the input mappings for a MIDI preset.
+ */
```
What's the reason for this?

daschuer replied:
Because it is hard to temporary disable portions of code via the block comments.

Holzhaus replied:
`#if 0 .... #endif` works but okay, i can remove them

Holzhaus replied:
Done.

Holzhaus committed:
Merge branch 'master' of github.com:mixxxdj/mixxx into controller-fixes-part2

Holzhaus committed:
controllers/dlgprefcontroller: Add edited suffix to filenames

    This allows editing user controller scripts (e.g. downloaded
    from the forum) as well.

Be-ing commented on src/controllers/dlgprefcontroller.cpp:
```diff
@@ -462,13 +462,24 @@ void DlgPrefController::savePreset() {
         return;
     }
 
-    if (m_pPreset->filePath().startsWith(resourcePresetsPath(m_pConfig))) {
```
Oh good, I was just going to ask you to append "(edited)" to edited presets from the user folder too.

Holzhaus committed:
controllers/dlgprefcontroller: Tick Enabled checkbox on preset change

Holzhaus committed:
controllers/dlgprefcontroller: Display "No scripts" for scriptless presets

Holzhaus committed:
controllers: Fix comments and include guards

Holzhaus commented:
There's still a lot that could be done to improve handling of controllers in Mixxx, but I consider this PR ready to merge. Please test.

Be-ing commented:
Great work! Thanks for cleaning up this mess :)

mevsme commented:
I still have a question. So what is the workflow if I want to modify the mapping? Do it in the mixxx's controllers folder?

ronso0 commented:
As soon as you change a control in Pref > Controllers > YourController > In/Output mapppings the XML is copied to the controllers folder. (js mapping still has to be copied there manually)

If you want change a mapping for a PR you can simply work in the git res/controllers folder.
